SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialSpanIdCache.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+
+#include "Misc/DateTime.h"
+#include "SpatialView/EntityComponentId.h"
+
+DECLARE_LOG_CATEGORY_EXTERN(LogSpatialSpanIdStore, Log, All);
+
+namespace SpatialGDK
+{
+class SpatialSpanIdCache
+{
+public:
+	struct FieldSpanIdUpdate
+	{
+		uint32 FieldId;
+		worker::c::Trace_SpanId NewSpanId;
+		worker::c::Trace_SpanId OldSpanId;
+	};
+
+	void ComponentAdd(const Worker_Op& Op);
+	bool ComponentRemove(const Worker_Op& Op);
+	TArray<FieldSpanIdUpdate> ComponentUpdate(const Worker_Op& Op);
+
+	void AddSpanId(const EntityComponentId& Id, const uint32 FieldId, worker::c::Trace_SpanId SpanId);
+	bool DropSpanId(const EntityComponentId& Id, const uint32 FieldId);
+	bool DropSpanIds(const EntityComponentId& Id);
+
+	bool GetSpanId(const EntityComponentId& Id, const uint32 FieldId, worker::c::Trace_SpanId& OutSpanId, bool bRemove = true);
+	bool GetMostRecentSpanId(const EntityComponentId& Id, worker::c::Trace_SpanId& OutSpanId, bool bRemove = true);
```

Comment:
I think this class is meant to just be a map from entity component to span ID. If it needs to be more complicated than that I think it needs a comment exaplining why.

IIRC last time this was discussed it was something like a map from EC -> Span
When you add a component you add to the map
when you update you get a new span ID caused by whatever was in the map + the new update
when you remove a component you remove it from the map.

You probably need to update this when sending stuff too. It's basically a mirror view and it might be something that we have to put into the view at some point. Although I don't really want to integrate it more into the view delta creation than it has to be.

Reply:
It maps Entity Component Field Id to SpanId. I assume the a component update only contains deltas for changes fields. That means you could receive two updates to the same component modifying different fields. If we are only concerned with "what is the last spanId for a componet that got update" rather than "what is the last spanId for a componet field that got update" we need some kind of temporal data.

TBH We could have another map from Entity Component to SpanId as well for when you don't care about fieldId.

Reply:
The main question I have on this is how much do we actually care about field ID spans vs component spans? As the latter sounds expensive although I could see it being useful for RPCs.

Reply:
Thinking about it a bit I don't think you can actually use field ID information reliably. I think in the rpc case it's not actually enough, you'd still need stuff in schema. So I think I'd recommend not bothering.